README.md
# add-contributors

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/b8d0af034c5c4699805c6aca898787e7)](https://app.codacy.com/manual/bobankhshen/add-contributors?utm_source=github.com&utm_medium=referral&utm_content=BobAnkh/add-contributors&utm_campaign=Badge_Grade_Dashboard)
![GitHub release (latest by date)](https://img.shields.io/github/v/release/BobAnkh/add-contributors?color=orange&logo=github-actions)
![language-python](https://img.shields.io/github/languages/top/BobAnkh/add-contributors?logo=python&logoColor=yellow)
![LICENSE Apache-2.0](https://img.shields.io/github/license/BobAnkh/add-contributors?logo=apache)

A Github Action to add contributors to your markdown file(i.e. README.md) automatically on schedule or triggered by events

Specifically handle unreachable Chinese context (着重解决了中文内容乱码的问题)

Feel free to submit a pull request or an issue, but make sure to follow the templates

Welcome contributors to improve this project together!

## Usage

Create a workflow file such as `.github/workflows/contributors.yml` (you can find it in this repo)

```yaml
name: Add contributors
on:
  schedule:
    - cron:  '20 20 * * *'
# push:
#   branches:
#     - master

jobs:
  add-contributors:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - uses: BobAnkh/add-contributors@master
      with:
        CONTRIBUTOR: '### Contributors'
        COLUMN_PER_ROW: '6'
        ACCESS_TOKEN: ${{secrets.GITHUB_TOKEN}}
        IMG_WIDTH: '100'
        FONT_SIZE: '14'
        PATH: '/README.md'
        COMMIT_MESSAGE: 'docs(README): update contributors'
        AVATAR_SHAPE: 'round'
```

### Parameters

| Parameter            | Description                                                 | Required | Default                                            |
| -------------------- | ----------------------------------------------------------- | -------- | -------------------------------------------------- |
| REPO_NAME            | Repository name                                             | no       | `''` which means current repository                |
| CONTRIBUTOR          | Where you want to add contributors list                     | no       | `### Contributors`                                 |
| COLUMN_PER_ROW       | Number of contributors per row                              | no       | `6`                                                |
| ACCESS_TOKEN         | Github Access Token                                         | yes      | You can just pass `${{secrets.GITHUB_TOKEN}}`      |
| IMG_WIDTH            | Width of img                                                | no       | `100`                                              |
| FONT_SIZE            | Font size of name (px)                                      | no       | `14`                                               |
| PATH                 | Path to the file you want to add contributors' list         | no       | `/README.md`                                       |
| BRANCH               | The branch to update file specified in PATH                 | no       | `''` which means default branch                    |
| PULL_REQUEST         | Open a new pull request if set to a target branch name      | no       | `''` which means not open pull request by default  |
| COMMIT_MESSAGE       | commit message                                              | no       | `docs(README): update contributors`                |
| AVATAR_SHAPE         | Set `round` for round avatar and `square` for square avatar | no       | square                                             |
| IGNORED_CONTRIBUTORS | Ignored contributors, seperated by comma                    | no       | `''`                                               |

> NOTE: You should leave a blank line after the `CONTRIBUTOR` line for the first time
>
> NOTE: Github seems not support image style in markdown file rendering yet
>
> NOTE: `IGNORED_CONTRIBUTORS` takes **display name** not **username**
> 
> NOTE: `PULL_REQUEST` must be used with `BRANCH` together, both **should be provided** if you want to **open a pull request**

## Maintainer

[@BobAnkh](https://github.com/BobAnkh)

## How to contribute

You should follow our [Code of Conduct](/CODE_OF_CONDUCT.md).

See [CONTRIBUTING GUIDELINES](/CONTRIBUTING.md) for contributing conventions.

Make sure to pass all the tests before submitting your code. You can conduct `pytest -ra` at the root directory to run all tests.

You can use local mode when develope it on your local machine, here is the command-line help info:

```console
usage: main.py [-h] [-m MODE] [-f FILE] [-o OUTPUT] [-t TOKEN]

optional arguments:
  -h, --help            show this help message and exit
  -m MODE, --mode MODE  choose to use local-dev mode or on github action mode.
                        Valid values are 'local' or 'github'
  -f FILE, --file FILE  configuration file to read from when running local-dev
                        mode
  -o OUTPUT, --output OUTPUT
                        output file when running local-dev mode
  -t TOKEN, --token TOKEN
                        Github Access Token
```

### Contributors

<table>
<tr>
    <td align="center" style="word-wrap: break-word; width: 150.0; height: 150.0">
        <a href=https://github.com/BobAnkh>
            <img src=https://avatars.githubusercontent.com/u/44333669?v=4 width="100;"  style="border-radius:50%;align-items:center;justify-content:center;overflow:hidden;padding-top:10px" alt=BobAnkh/>
            <br />
            <sub style="font-size:14px"><b>BobAnkh</b></sub>
        </a>
    </td>
    <td align="center" style="word-wrap: break-word; width: 150.0; height: 150.0">
        <a href=https://github.com/baileythegreen>
            <img src=https://avatars.githubusercontent.com/u/12277715?v=4 width="100;"  style="border-radius:50%;align-items:center;justify-content:center;overflow:hidden;padding-top:10px" alt=Bailey Harrington/>
            <br />
            <sub style="font-size:14px"><b>Bailey Harrington</b></sub>
        </a>
    </td>
    <td align="center" style="word-wrap: break-word; width: 150.0; height: 150.0">
        <a href=https://github.com/casperklein>
            <img src=https://avatars.githubusercontent.com/u/590174?v=4 width="100;"  style="border-radius:50%;align-items:center;justify-content:center;overflow:hidden;padding-top:10px" alt=Casper/>
            <br />
            <sub style="font-size:14px"><b>Casper</b></sub>
        </a>
    </td>
</tr>
</table>

## LICENSE

[Apache-2.0](/LICENSE) © BobAnkh


CHANGELOG.md
# CHANGELOG

## [v0.2.2](https://github.com/BobAnkh/add-contributors/releases/tag/v0.2.2) - 2021-09-25 02:39:12

New Feature:

- Split param `PATH` into `PATH` and `BRANCH`: you can specify which file path in `PATH` param and which branch to update changes in `BRANCH` param.
- Add new param `PULL_REQUEST`, which must be used together with param `BRANCH`. Default set to `''` means not to open a pull request. You can set a target branch name in `PULL_REQUEST` that is different with `BRANCH` so that to do the following things:
  - Commit changes(if have) to file specified in `PATH` in `BRANCH`(if not set means default branch)
  - If `PULL_REQUEST` set to a different branch with `BRANCH` then a new pull request will be opened.

### Feature

- pull:
  - add new param for pull request ([6a63a21](https://github.com/BobAnkh/add-contributors/commit/6a63a219eb27541846688e999539d2f70df818e5))

- param:
  - add repo detect and BRANCH option ([74cbba3](https://github.com/BobAnkh/add-contributors/commit/74cbba39759181b381edd659674c5ab0d65e3bc8))

### Bug Fixes

- main:
  - resolve create_pull func call (#50) ([70fb0e5](https://github.com/BobAnkh/add-contributors/commit/70fb0e516dcf18549ef503dd6a8d6295f5a1dc19)) ([#50](https://github.com/BobAnkh/add-contributors/pull/50))
  - resolve `create_pull` func call ([db74794](https://github.com/BobAnkh/add-contributors/commit/db747947e2a4135486cfc34ec02da07016f240ce))

### Documentation

- readme:
  - update usage of new param ([de6c43a](https://github.com/BobAnkh/add-contributors/commit/de6c43a0c154b093a210d8bdf7ab283aba8452dd))
  - update usage for new param ([57aa424](https://github.com/BobAnkh/add-contributors/commit/57aa42418b403c07faf93b700d2349c0231cc53f))
  - update usage for params ([889f1c9](https://github.com/BobAnkh/add-contributors/commit/889f1c92705c63af28354e01dac9349607526960))

- README:
  - update contributors ([b4ec2db](https://github.com/BobAnkh/add-contributors/commit/b4ec2dbcd07093b997b9f6ace9bb5a467b93339c))

## [v0.2.1](https://github.com/BobAnkh/add-contributors/releases/tag/v0.2.1) - 2021-07-17 03:39:06

Fix some small bugs:
- now can specify debug config file in args
- now can exclude no-name accounts

### Bug Fixes

- contributor:
  - fix no name accounts' exclusion ([73c1109](https://github.com/BobAnkh/add-contributors/commit/73c1109d057357e2a76e71fe00db880f02c0ff5d))

- local:
  - add args-specify config file ([774a109](https://github.com/BobAnkh/add-contributors/commit/774a10968205067ec224f5fce2e4c28b9f83d1bb))

### Documentation

- readme:
  - update usage note of attributes ([1f2cd9f](https://github.com/BobAnkh/add-contributors/commit/1f2cd9f3e95509ba5437f0f028915d1792782eb1))

## [v0.2.0](https://github.com/BobAnkh/add-contributors/releases/tag/v0.2.0) - 2021-07-16 06:47:07

Add a new feature proposed in #46 to exclude certain contributors(e.g.  @renovate-bot and @codacy-badger)

Resolves: #46 

### Feature

- contributor:
  - support to ignore ocntributors ([d049565](https://github.com/BobAnkh/add-contributors/commit/d0495656fe47ec6cec1837b7bcf0a2b953ed9b92))

### Bug Fixes

- local:
  - resolve token read from file in local-dev mode ([5621757](https://github.com/BobAnkh/add-contributors/commit/5621757f674df38b746d57711f6d44a52442e23f))

### Documentation

- contributor:
  - update params usage ([2acb218](https://github.com/BobAnkh/add-contributors/commit/2acb2184554d1984a2e956f809b7490121cc980d))

- README:
  - update contributors ([969bc58](https://github.com/BobAnkh/add-contributors/commit/969bc58d3e0e3b6e71083ef0c6631ff1fa696750))
  - add info for tests ([6647d3d](https://github.com/BobAnkh/add-contributors/commit/6647d3d22547d2a5865e2b3d1bf24e92f076b5fa))

### Performance Improvements

- Dockerfile:
  - merge layers ([1f8ecc6](https://github.com/BobAnkh/add-contributors/commit/1f8ecc6677001e72a8498a37b2c4d04d59a5587b))

## [v0.1.0](https://github.com/BobAnkh/add-contributors/releases/tag/v0.1.0) - 2021-01-09 04:21:08

Refactor the data structure. No breaking changes. Prepare to add some tests

Add local mode for developing on local machine.

### Feature

- mode:
  - add local-dev mode ([c4e8115](https://github.com/BobAnkh/add-contributors/commit/c4e8115850f4148323689371463396d8dd6e272a)) ([#36](https://github.com/BobAnkh/add-contributors/pull/36))

### Documentation

- README:
  - update info for local mode ([3f02f72](https://github.com/BobAnkh/add-contributors/commit/3f02f720122e6f9015e5cdaeaea8beac18cef2e4))
  - fix a typo ([90eedc1](https://github.com/BobAnkh/add-contributors/commit/90eedc1d8dae57ef168856d08ccf9758b23bbef2))

### Refactor

- main:
  - uncouple data and api to github ([27b5e5b](https://github.com/BobAnkh/add-contributors/commit/27b5e5b55a0b8134a4ebc2073b3267f06154a08c))

### Performance Improvements

- main:
  - use yaml.safe_load and add docstring ([44553f6](https://github.com/BobAnkh/add-contributors/commit/44553f6d59769836061740bebdecd422493bbd5e))

- data:
  - imporve data structure ([7098489](https://github.com/BobAnkh/add-contributors/commit/7098489d03873c0235edf349b24679f33400cc03))

## [v0.0.8](https://github.com/BobAnkh/add-contributors/releases/tag/v0.0.8) - 2020-12-12 00:47:26

Change layout of contributors table to make it more beautiful.

### Feature

- table-style:
  - change layout of contributors table (#34) ([54e695f](https://github.com/BobAnkh/add-contributors/commit/54e695f9222907293a2d42c98937678763518ce4)) ([#34](https://github.com/BobAnkh/add-contributors/pull/34))

### Bug Fixes

- locate:
  - change locate feature ([ef71274](https://github.com/BobAnkh/add-contributors/commit/ef71274049fccc95376b272ca95cacb32eaed9b8))

### Documentation

- README:
  - update contributors ([f229687](https://github.com/BobAnkh/add-contributors/commit/f2296872426a5a36168cdf73691035a3da5df4f1))

## [v0.0.7](https://github.com/BobAnkh/add-contributors/releases/tag/v0.0.7) - 2021-07-16 11:29:55

Support other branches.


tests/test_main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# @Author       : BobAnkh
# @Github       : https://github.com/BobAnkh
# @Date         : 2021-01-09 11:53:15
# @LastEditTime : 2021-01-10 22:01:16
# @Description  : Tests for main.py

import json
import random
import string
import sys

import pytest

sys.path.append('.')
import main

case = json.load(open('tests/case.json'))


@pytest.fixture(params=case['test_env_case'])
def generate_env(request):
    output = {}
    output['env_name'] = request.param
    output['env_value'] = ''.join(
        random.sample(string.ascii_letters + string.digits, 20))
    return output


def test_env(generate_env):
    env_name = generate_env['env_name']
    env_value = generate_env['env_value']
    main.set_local_env(env_name, env_value)
    msg = "Test case %s is wrong" % (env_name)
    assert env_value == main.get_inputs(env_name), msg


@pytest.mark.parametrize(
    'contributors_data, row, width, font_size, head,tail, rsb, contributors_table',
    case['test_table_case'])
def test_table(contributors_data, row, width, font_size, head, tail, rsb,
               contributors_table):
    assert contributors_table == main.generate_contributors_table(
        contributors_data, row, width, font_size, head, tail, rsb)


@pytest.mark.parametrize(
    'content, contributors_table, CONTRIBUTOR, PATH, text',
    case['test_content_case'])
def test_content(content, contributors_table, CONTRIBUTOR, PATH, text):
    assert text == main.generate_content(content, contributors_table,
                                         CONTRIBUTOR, PATH)


main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# @Author       : BobAnkh
# @Github       : https://github.com/BobAnkh
# @Date         : 2020-07-29 00:12:39
# @LastEditors  : BobAnkh
# @LastEditTime : 2021-09-24 16:28:29
# @FilePath     : /add-contributors/main.py
# @Description  : Main script of Github Action
# @Copyright 2020 BobAnkh

import argparse
import base64
import os
import re

import github
import yaml

head = '''<table>
<tr>'''
tail = '''
</tr>
</table>'''


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '-m',
        '--mode',
        help=
        'choose to use local-dev mode or on github action mode. Valid values are \'local\' or \'github\'',
        default='github')
    parser.add_argument(
        '-f',
        '--file',
        help='configuration file to read from when running local-dev mode',
        default='.github/workflows/contributors.yml')
    parser.add_argument('-o',
                        '--output',
                        help='output file when running local-dev mode',
                        default='local-dev.md')
    parser.add_argument('-t', '--token', help='Github Access Token')
    args = parser.parse_args()
    return args


class GithubContributors:
    '''
    Class for data interface of Github

    Use it to get contributors data and file content from Github and write new file content to Github
    '''
    def __init__(self, ACCESS_TOKEN, REPO_NAME, PATH, BRANCH, PULL_REQUEST, COMMIT_MESSAGE, IGNORED_CONTRIBUTORS):
        '''
        Initial GithubContributors

        Args:
            ACCESS_TOKEN (str): Personal Access Token for Github
            REPO_NAME (str): The name of the repository
            PATH (str): The path to the file
            BRANCH (str): The branch of the file
            PULL_REQUEST (str): Pull request target branch, none means do not open a pull request
            COMMIT_MESSAGE (str): Commit message you want to use
        '''
        self.COMMIT_MESSAGE = COMMIT_MESSAGE
        self.PATH = PATH
        self.BRANCH = BRANCH
        self.PULL_REQUEST = PULL_REQUEST
        self.IGNORED_CONTRIBUTORS = IGNORED_CONTRIBUTORS
        self.SHA = ''
        self.contributors_data = []
        self.file_content = ''
        # Use PyGithub to login to the repository
        # References: https://pygithub.readthedocs.io/en/latest/github_objects/Repository.html#github.Repository.Repository
        g = github.Github(ACCESS_TOKEN)
        self.repo = g.get_repo(REPO_NAME)

    def get_data(self):
        # get contributors data
        contributors = self.repo.get_contributors()
        print('[DEBUG]All contributors\' names:')
        for contributor in contributors:
            name = contributor.name
            avatar_url = contributor.avatar_url
            html_url = contributor.html_url
            if name == None:
                name = html_url[19:]
            print(f'[DEBUG]{name}:{html_url}')
            if name in self.IGNORED_CONTRIBUTORS or re.match('apps/', name):
                continue
            self.contributors_data.append({
                'name': name,
                'avatar_url': avatar_url,
                'html_url': html_url
            })
        # get file content
        contents = self.repo.get_contents(self.PATH, self.BRANCH)
        self.PATH = contents.path
        self.SHA = contents.sha
        base = contents.content
        base = base.replace('\n', '')
        self.file_content = base64.b64decode(base).decode('utf-8')

    def write_data(self, content):
        if content == self.file_content:
            pass
        else:
            self.repo.update_file(self.PATH, self.COMMIT_MESSAGE, content,
                                  self.SHA, self.BRANCH)
            print(f'[DEBUG] BRANCH: {self.BRANCH}, PULL_REQUEST: {self.PULL_REQUEST}')
            if self.PULL_REQUEST != '' and self.PULL_REQUEST != self.BRANCH:
                self.repo.create_pull(title=self.COMMIT_MESSAGE, body=self.COMMIT_MESSAGE, base=self.PULL_REQUEST, head=self.BRANCH, draft=False, maintainer_can_modify=True)

    def read_contributors(self):
        return self.contributors_data

    def read_file_content(self):
        return self.file_content


def set_local_env(env_name: str, env_value: str, prefix='INPUT'):
    '''
    set local env for dev

    Args:
        env_name (str): local env name.
        env_value (str): value of local env name.
        prefix (str, optional): prefix of env variable. Defaults to 'INPUT'.
    '''
    os.environ[prefix + '_{}'.format(env_name).upper()] = env_value


def get_inputs(input_name: str, prefix='INPUT') -> str:
    '''
    Get a Github actions input by name

    Args:
        input_name (str): input_name in workflow file.
        prefix (str, optional): prefix of input variable. Defaults to 'INPUT'.

    Returns:
        str: action_input

    References
    ----------
    [1] https://help.github.com/en/actions/automating-your-workflow-with-github-actions/metadata-syntax-for-github-actions#example
    '''
    return os.getenv(prefix + '_{}'.format(input_name).upper())


def generate_contributors_table(contributors_data, COLUMN_PER_ROW, img_width,
                                font_size, head_format, tail_format, shape):
    '''
    Generate the contributors table in html format using a given template

    Args:
        contributors_data (list): a list of dict which contains contributors' name, avatar_url and html_url
        COLUMN_PER_ROW (int): number of contributors per row
        img_width (int): width of img
        font_size (int): font size of name
        head_format (str): html_format for table head
        tail_format (str): html_format for table tail
        shape (str): round for round avatar and square for square avatar

    Returns:
        str: contributors table in html format
    '''
    USER = 0
    HEAD = head_format
    TAIL = tail_format
    cell_width = 1.5 * img_width
    cell_height = 1.5 * img_width
    for contributor in contributors_data:
        name = contributor['name']
        avatar_url = contributor['avatar_url']
        html_url = contributor['html_url']
        if USER >= COLUMN_PER_ROW:
            new_tr = '''\n</tr>\n<tr>'''
            HEAD = HEAD + new_tr
            USER = 0
        if shape == 'round':
            img_style = ' style="border-radius:50%;align-items:center;justify-content:center;overflow:hidden;padding-top:10px"'
        else:
            img_style = ''
        td = f'''
    <td align="center" style="word-wrap: break-word; width: {cell_width}; height: {cell_height}">
        <a href={html_url}>
            <img src={avatar_url} width="{img_width};" {img_style} alt={name}/>
            <br />
            <sub style="font-size:{font_size}px"><b>{name}</b></sub>
        </a>
    </td>'''
        HEAD = HEAD + td
        USER += 1
    HEAD = HEAD + TAIL
    return HEAD


def generate_content(file_content, contributors_table, CONTRIBUTOR, PATH):
    '''
    Generate the whole content with contributors table

    Args:
        file_content (str): content of target file
        contributors_table (str): contributors list
        CONTRIBUTOR (str): where you want to write the contributors list
        PATH (str): the file to write

    Raises:
        Exception: the target file does not have the CONTRIBUTOR section

    Returns:
        str: the whole content with contributors table
    '''
    text = file_content
    text_str = text.split(CONTRIBUTOR)
    if len(text_str) == 1:
        print('[DEBUG]: ', text, '\n[DEBUG]')
        raise Exception("File '" + PATH + "' does not have '" + CONTRIBUTOR +
                        "' section")
    if re.match(r'\n+', text_str[1]):
        lf_num = re.match(r'\n+', text_str[1]).span()[1]
        text_str[1] = text_str[1][lf_num:]
    else:
        lf_num = 0
        print('[DEBUG-lr_num]: ', text_str[1])
    if re.match(head, text_str[1]):
        end = text_str[1].split(tail)
        end[0] = end[0] + tail
    else:
        end = ['', '\n' * (lf_num + 1) + text_str[1]]
    end[0] = contributors_table
    text = text_str[0] + CONTRIBUTOR + '\n' + end[0] + end[1]
    return text


def set_env_from_file(file, args, prefix='INPUT'):
    '''
    Set env when use local-dev mode

    Args:
        file (str): path to config file
        args (object): cmdline argument
        prefix (str, optional): prefix of env. Defaults to 'INPUT'.
    '''
    f = open(file, encoding='utf-8')
    y = yaml.safe_load(f)
    for job in y['jobs'].values():
        for step in job['steps']:
            if re.match(r'BobAnkh/add-contributors', step['uses']):
                params = step['with']
                break
    option_params = [
        'REPO_NAME', 'CONTRIBUTOR', 'COLUMN_PER_ROW', 'ACCESS_TOKEN',
        'IMG_WIDTH', 'FONT_SIZE', 'PATH', 'COMMIT_MESSAGE', 'AVATAR_SHAPE', 'IGNORED_CONTRIBUTORS'
    ]
    for param in option_params:
        if param not in params.keys():
            if param == 'ACCESS_TOKEN' and args.token:
                tmp = args.token
            else:
                tmp = input('Please input the value of ' + param + ':')
        elif param == 'ACCESS_TOKEN':
            if re.match(r'\$\{\{secrets\.', params[param]):
                if args.token:
                    tmp = args.token
                else:
                    tmp = input('Please input the value of ' + param + ':')
            else:
                tmp = params[param]
        elif param == 'REPO_NAME' and params[param] == '':
            tmp = input('Please input the value of ' + param + ':')
        else:
            tmp = params[param]
        set_local_env(param, tmp, prefix)


def main():
    args = argument_parser()
    if args.mode == 'local':
        set_env_from_file(args.file, args)
    elif args.mode == 'github':
        pass
    else:
        print("Illegal mode option, please type \'-h\' to read the help")
        os.exit()
    ACCESS_TOKEN = get_inputs('ACCESS_TOKEN')
    REPO_NAME = get_inputs('REPO_NAME')
    if REPO_NAME == '':
        REPO_NAME = get_inputs('REPOSITORY', 'GITHUB')
    CONTRIBUTOR = get_inputs('CONTRIBUTOR') + '\n'
    COLUMN_PER_ROW = int(get_inputs('COLUMN_PER_ROW'))
    IMG_WIDTH = int(get_inputs('IMG_WIDTH'))
    FONT_SIZE = int(get_inputs('FONT_SIZE'))
    PATH = get_inputs('PATH')
    BRANCH = get_inputs('BRANCH')
    if BRANCH == '':
        BRANCH = github.GithubObject.NotSet
    PULL_REQUEST = get_inputs('PULL_REQUEST')
    COMMIT_MESSAGE = get_inputs('COMMIT_MESSAGE')
    AVATAR_SHAPE = get_inputs('AVATAR_SHAPE')
    IGNORED_CONTRIBUTORS = [x.lstrip().rstrip() for x in get_inputs('IGNORED_CONTRIBUTORS').split(',')]
    Contributors = GithubContributors(ACCESS_TOKEN, REPO_NAME, PATH, BRANCH, PULL_REQUEST,
                                      COMMIT_MESSAGE, IGNORED_CONTRIBUTORS)
    Contributors.get_data()
    contributors_table = generate_contributors_table(
        Contributors.read_contributors(), COLUMN_PER_ROW, IMG_WIDTH, FONT_SIZE,
        head, tail, AVATAR_SHAPE)
    content = generate_content(Contributors.read_file_content(),
                               contributors_table, CONTRIBUTOR, PATH)
    if args.mode == 'local':
        with open(args.output, 'w', encoding='utf-8') as f:
            f.write(content)
    else:
        Contributors.write_data(content)


if __name__ == '__main__':
    main()
